packages/transaction-controller/src/helpers/ResimulateHelper.ts

```diff
@@ -1,31 +1,139 @@
+const log = createModuleLogger(projectLogger, 'resimulate-helper');
 
 export const RESIMULATE_PARAMS = ['to', 'value', 'data'] as const;
 export const BLOCKAID_RESULT_TYPE_MALICIOUS = 'Malicious';
 export const VALUE_COMPARISON_PERCENT_THRESHOLD = 5;
 export const BLOCK_TIME_ADDITIONAL_SECONDS = 60;
+export const RESIMULATE_INTERVAL_MS = 3000;
 
 export type ResimulateResponse = {
   blockTime?: number;
   resimulate: boolean;
 };
 
+export type ResimulateHelperOptions = {
+  getTransactions: () => TransactionMeta[];
+  onStateChange: (listener: () => void) => void;
+  updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+};
+
+export class ResimulateHelper {
+  // Map of transactionId <=> isActive
+  readonly #activeResimulations: Map<string, boolean> = new Map();
+
+  // Map of transactionId <=> intervalId
+  readonly #intervalIds: Map<string, NodeJS.Timeout> = new Map();
+
+  readonly #getTransactions: () => TransactionMeta[];
+
+  readonly #updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+
+  constructor({
+    getTransactions,
+    updateSimulationData,
+    onStateChange,
+  }: ResimulateHelperOptions) {
+    this.#getTransactions = getTransactions;
+    this.#updateSimulationData = updateSimulationData;
+
+    onStateChange(() => {
+      const unapprovedTransactions = this.#getUnapprovedTransactions();
+      const unapprovedTransactionIds = new Set(
+        unapprovedTransactions.map((tx) => tx.id),
+      );
+
+      // Start or stop resimulation based on the current isFocused state
+      unapprovedTransactions.forEach((transactionMeta) => {
+        if (transactionMeta.isFocused) {
+          this.#start(transactionMeta);
+        } else {
+          this.#stop(transactionMeta);
+        }
+      });
+
+      // Force stop any running active resimulation that are no longer unapproved transactions list
+      this.#activeResimulations.forEach((isActive, transactionId) => {
+        if (isActive && !unapprovedTransactionIds.has(transactionId)) {
+          this.#stop({
+            id: transactionId,
+            // Forcing this to false to ensure the resimulation is stopped
+            isFocused: false,
+          } as unknown as TransactionMeta);
+        }
+      });
+    });
+  }
+
+  #start(transactionMeta: TransactionMeta) {
+    const { id: transactionId } = transactionMeta;
+    const isActive = this.#activeResimulations.get(transactionId);
+    if (!transactionMeta.isFocused || isActive) {
+      return;
+    }
+
+    const listener = async () => {
+      try {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.#updateSimulationData(transactionMeta);
+      } catch (error) {
+        /* istanbul ignore next */
+        log('Error during transaction resimulation', error);
+      }
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    const intervalId = setInterval(listener, RESIMULATE_INTERVAL_MS);
```

Comment:
Should we use a `setTimeout` chain rather than `setInterval` otherwise we risk slow simulations overlapping as we don't use a mutex?